test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,878 @@
+        if (elementInfo == null) {
+            error("Element :" + id + " was not mapped");
+            return;
+        }
+        String version = elementInfo.introducedStable;
+        if (moduleVersion == null) {
+            error("Unable to retrieve `@since` for " + elementCategory + id);
+        } else {
+            checkEquals(moduleVersion, version, id);
+        }
+    }
+
+    private String getModuleVersionFromFile(Path moduleDirectory) {
+        Path moduleInfoFile = moduleDirectory.resolve("module-info.java");
+        String version = null;
+        if (Files.exists(moduleInfoFile)) {
+            try {
+                String moduleInfoContent = Files.readString(moduleInfoFile);
+                version = extractSinceVersionFromText(moduleInfoContent).toString();
+            } catch (IOException e) {
+                error("module-info.java not found or couldn't be opened AND this module has no unqualified exports");
+            } catch (NullPointerException e) {
+                error("module-info.java does not contain an `@since`");
+            }
+        }
+        return version;
+    }
+
+    private String getPackageVersionFromFile(Path moduleDirectory, ModuleElement.ExportsDirective ed) {
+        Path pkgInfo = moduleDirectory.resolve(ed.getPackage()
+                        .getQualifiedName()
+                        .toString()
+                        .replace(".", File.separator)
+                )
+                .resolve("package-info.java");
+
+        String packageTopVersion = null;
+        if (Files.exists(pkgInfo)) {
+            try {
+                String packageContent = Files.readString(pkgInfo);
+                packageTopVersion = extractSinceVersionFromText(packageContent).toString();
+            } catch (IOException e) {
+                error(ed.getPackage().getQualifiedName() + ": package-info.java not found or couldn't be opened");
+            } catch (NullPointerException e) {
+                error(ed.getPackage().getQualifiedName() + ": package-info.java  does not contain an `@since`");
+            }
+        }
+        return packageTopVersion;
+    }
+
+
+    private void analyzePackageCheck(PackageElement pe, JavacTask ct, EffectiveSourceSinceHelper javadocHelper) {
+        List<TypeElement> typeElements = ElementFilter.typesIn(pe.getEnclosedElements());
+        for (TypeElement te : typeElements) {
+            analyzeClassCheck(te, null, javadocHelper, ct.getTypes(), ct.getElements());
+        }
+    }
+
+    private void analyzeClassCheck(TypeElement te, String version, EffectiveSourceSinceHelper javadocHelper,
+                                   Types types, Elements elementUtils) {
+        String currentjdkVersion = String.valueOf(Runtime.version().feature());
+        if (!isDocumented(te)) {
+            return;
+        }
+        checkElement(te.getEnclosingElement(), te, types, javadocHelper, version, elementUtils);
+        if( te.getKind() == ElementKind.RECORD){
```

Comment:
This seems a bit too broad. Ignoring all members of a record type may lead to missed members. `Elements.getOrigin` may be unusable here, but I wonder what exactly is the problem which this condition is trying to solve?

Reply:
I changed my approach in [ac4df85](https://github.com/openjdk/jdk/pull/18934/commits/ac4df85615557306da606a712ee02e7155e5412f). I beleieve it is a bit clearer now. I am only skipping some very common methods (`toString`, `equals` and `hashCode`)

Previously, I failed to get the effective `@since` of some common methods eclosed within a record.
I will use this as an example
https://github.com/openjdk/jdk/blob/b92bd671835c37cff58e2cdcecd0fe4277557d7f/src/jdk.jshell/share/classes/jdk/jshell/execution/JdiDefaultExecutionControl.java#L346

`te.getEnclosedElements()` on this record returns the following (I truncated `jdk.jshell.execution.JdiDefaultExecutionControl.JdiStarter.TargetDescription` into `TargetDescription`)
```
method: void TargetDescription.<init>(com.sun.jdi.VirtualMachine,java.lang.Process) 
method: void TargetDescription.wait(long,int) 
method: java.lang.Process TargetDescription.process() 
method: com.sun.jdi.VirtualMachine TargetDescription.vm() 
method: java.lang.String TargetDescription.toString() 
method: int TargetDescription.hashCode() 
method: java.lang.Object TargetDescription.clone() 
method: java.lang.Class TargetDescription.getClass() 
method: void TargetDescription.finalize() 
method: void TargetDescription.notify() 
method: void TargetDescription.wait(long) 
method: boolean TargetDescription.equals(java.lang.Object) 
method: void TargetDescription.notifyAll() 
method: void TargetDescription.wait() 
```

Getting the effective `@since` for these 5 methods in particular would fail
```
method: java.lang.String TargetDescription.toString():
method: int TargetDescription.hashCode(): 
method: boolean TargetDescription.equals(java.lang.Object):
method: com.sun.jdi.VirtualMachine TargetDescription.vm(): 
method: java.lang.Process TargetDescription.process():
```
It seemed to me that these methods would be added in the same version as the record, so I skipped them all.
I didn't think of a user adding a static variable in a later release and that it was too broad, my bad.